### PR TITLE
Ux 1658

### PR DESCRIPTION
**Pull request by bnlb**
Resolves [FRONTEND-1658](https://algorithmia.atlassian.net/browse/FRONTEND-1658)

@langlk has a branch in test right now, but I'll deploy this to test and confirm there aren't any surprises before merging. 

If you try to run `node server/index.js` you'll get an error, as the `sites/` directory won't exist locally. You'll have to run this inside of Docker to see it in action:

```
docker build -t dev-center:latest .
docker run -p 3000:3000 dev-center:latest
```

Visit http://localhost:3000/developers

### CODE REVIEW

**Comment by langlk** [comment]
> Local render looks good except for head/foot (TBD via merge l8r)
> 
> Got some 404s on the server tho:
> [2019-09-27 12:35:46] ERROR `/developers/synapse/dist/scoped-synapse.css' not found. [2019-09-27 12:35:47] ERROR `/webapi/user' not found.
> [2019-09-27 12:35:47] ERROR `/favicon.ico' not found.

For the synapse error, you'll want to make sure the submodule is updated and built. ([step 5 here](https://github.com/algorithmiaio/dev-center#getting-set-up))

**Comment by langlk** [comment]
Looks like the angular code on the getting started page isn't working?

**Comment by bnlb** [comment]
> Looks like the angular code on the getting started page isn't working?

As long as you're not seeing any console errors we should be good. The reason it's not working is because it makes a request to `/webapi/user` to get the user's info, but we're not actually proxying anything to the Play server when we start this Docker container, since that route would normally be handled by the web client's server code.

If you start the container using `-p 4000:3000` and then do `npm run localdevcenter` like normal you'll see it work just fine.

**Comment by langlk** [comment]
> > Looks like the angular code on the getting started page isn't working?
> 
> As long as you're not seeing any console errors we should be good. The reason it's not working is because it makes a request to `/webapi/user` to get the user's info, but we're not actually proxying anything to the Play server when we start this Docker container, since that route would normally be handled by the web client's server code.
> 
> If you start the container using `-p 4000:3000` and then do `npm run localdevcenter` like normal you'll see it work just fine.

Yes, but it shouldn't need to make the `webapi/user` call to work. The copying functionality and language selector should be fine without that.

**Comment by langlk** [comment]
> > > Looks like the angular code on the getting started page isn't working?
> > 
> > 
> > As long as you're not seeing any console errors we should be good. The reason it's not working is because it makes a request to `/webapi/user` to get the user's info, but we're not actually proxying anything to the Play server when we start this Docker container, since that route would normally be handled by the web client's server code.
> > If you start the container using `-p 4000:3000` and then do `npm run localdevcenter` like normal you'll see it work just fine.
> 
> Yes, but it shouldn't need to make the `webapi/user` call to work. The copying functionality and language selector should be fine without that.

I'm also seeing this console error:
```
core.cat.js:1 Error: [$controller:ctrlreg] The controller with the name 'GettingStartedControl' is not registered.
http://errors.angularjs.org/1.6.5/$controller/ctrlreg?p0=GettingStartedControl
    at core.cat.js:1
    at $controller (core.cat.js:1)
    at setupControllers (core.cat.js:1)
    at nodeLinkFn (core.cat.js:1)
    at compositeLinkFn (core.cat.js:1)
    at compositeLinkFn (core.cat.js:1)
    at compositeLinkFn (core.cat.js:1)
    at compositeLinkFn (core.cat.js:1)
    at compositeLinkFn (core.cat.js:1)
    at compositeLinkFn (core.cat.js:1)
(anonymous) @ core.cat.js:1
(anonymous) @ core.cat.js:1
$apply @ core.cat.js:1
bootstrapApply @ core.cat.js:1
invoke @ core.cat.js:1
doBootstrap @ core.cat.js:1
bootstrap @ core.cat.js:1
angularInit @ core.cat.js:1
(anonymous) @ core.cat.js:1
mightThrow @ core.cat.js:1
process @ core.cat.js:1
setTimeout (async)
(anonymous) @ core.cat.js:1
fire @ core.cat.js:1
fireWith @ core.cat.js:1
fire @ core.cat.js:1
fire @ core.cat.js:1
fireWith @ core.cat.js:1
ready @ core.cat.js:1
completed @ core.cat.js:1
```

**Comment by bnlb** [comment]
> > > Looks like the angular code on the getting started page isn't working?
> > 
> > 
> > As long as you're not seeing any console errors we should be good. The reason it's not working is because it makes a request to `/webapi/user` to get the user's info, but we're not actually proxying anything to the Play server when we start this Docker container, since that route would normally be handled by the web client's server code.
> > If you start the container using `-p 4000:3000` and then do `npm run localdevcenter` like normal you'll see it work just fine.
> 
> Yes, but it shouldn't need to make the `webapi/user` call to work. The copying functionality and language selector should be fine without that.

Ah, I misunderstood. I thought the credentials just weren't updating. I can reproduce.

**Comment by bnlb** [comment]
@langlk This is fixed. Angular required the `ng-app` directive to be present on the containing element, and since that was supplied by the web client repo, it broke when rendered directly.

I ran into an issue while fixing this where the dropdowns stopped working despite Angular running without any issues. It turns out Bootstrap was handling that logic, not Angular. @peckjon do you know of any other cases where there could be another implicit dependency that wouldn't result in a 404 but could still cause a silent failure?

With both Bootstrap and Angular running on the pages again we should be good to go. All of the interactive functionality I can find is working fine. Wanted your 2¢ before merging anything in case I'm forgetting something.

For what it's worth, this is live in test. (Even though the fix should be verified via that Docker command I posted above.)

**Comment by peckjon** [comment]
Not explicitly, but I’ll vet the live test version carefully

**Comment by peckjon** [comment]
LL we have some overflow issues on mobile: https://test.algorithmia.com/developers/integrations/event_listeners

Likely this is due to image widths not being constrained?

**Comment by bnlb** [comment]
This is fixed. @langlk recently added some flexbox styles to dev-center pages. Turns out our `<figure>` elements could overflow (images had a `max-width` applied so it only happened on pages with a `<figure>`) the width of the page, which is why we didn't see it on any other pages.

File: package.json [hunk 1]
```diff
@@ -6,7 +6,13 @@
   "author": "Algorithmia <mia@algorithmia.com>",
```
Comment (langlk) [inline package.json]: Could we also commit the `package-lock.json` changes?

File: README.md [hunk 2]
```diff
@@ -56,7 +56,7 @@ gem install --install-dir vendor/bundle/ruby/<ruby_version_numbwe> nokogiri -v "
  5. Run `git submodule init`, initializing the submodule for our UI toolkit, Synapse. **Do bear in mind that you will need to able to communicate with [GitHub via SSH](https://help.github.com/en/articles/connecting-to-github-with-ssh) for this to function as expected.**
  6. Run `git submodule update`, cloning down submodule code.
  7. To build Synapse dist files, run `cd synapse && npm ci && npm run build && cd ..`
- 7. Start the test server: `bundle exec jekyll serve` (append `--config _config.yml,_config-dev.yml` to use dev cdn config; to simulate an enterprise deploy, also add `_config-enterprise.yml`)
+ 8. Start the test server: `bundle exec jekyll serve` (append `--config _config.yml,_config-enterprise.yml` to simulate an enterprise deploy)
```
Comment (langlk) [inline README.md]: Could we also add instructions for running the node server?

File: package.json (hunk 1 above)
Comment (bnlb) [reply package.json]: Nice catch. This was in `.gitignore` for some reason. Updated.

File: README.md (hunk 2 above)
Comment (bnlb) [reply README.md]: Done!

File: README.md [hunk 3]
```diff
@@ -62,6 +62,18 @@ You can now see the developer center at <http://localhost:4000>.
 
 Note: the header and footer may not look right. This is inconsequential as they are stripped from the page before embedding into the webserver.
 
+### Running the Node server locally
+
+ 1. First, build the `./sites` folder to serve content out of:
+
+    ```
+    mkdir /sites && \
```
Comment (langlk) [inline README.md]: Is there any chance we could make this a script instead of requiring the copy-pasting?

**Comment by andryanmiller** [comment]
As discussed IRL, it's actually safe to leave the CDN prefix, even as we switch over to the new infra.

File: README.md (hunk 3 above)
Comment (bnlb) [reply README.md]: I like that. [Updated.](https://github.com/algorithmiaio/dev-center/pull/210/commits/f42c992d1cf779b7abc6f8444df2f13a6fe3290c)

**Comment by bnlb** [comment]
> As discussed IRL, it's actually safe to leave the CDN prefix, even as we switch over to the new infra.

@andryanmiller this is updated and currently deployed to test. Wanted to give you a heads up in case you wanted to take another look now that the PR is significantly smaller.

File: css/vendors.css [hunk 4]
```diff
@@ -12500,19 +12500,19 @@ fieldset[disabled] .btn-flat:hover {
 }
 
 .icon-check-gray {
-  background-image: url(/assets/images/icons/check_simple_gray.svg)
```
Comment (andryanmiller) [inline css/vendors.css]: 💭 What's the motivation for the URL changes here? Ideally, this app never needs to know the path it lives under. Could a `<base>` tag help us out?

Comment (bnlb) [reply css/vendors.css]: I got 404s while running the Node server locally for these assets. Turns out they live in the Play repo [here](https://github.com/algorithmiaio/algorithmia/tree/develop/public/images/icons). I wanted to ensure that any assets the Node server needs in order to render its pages are contained within this repo. (Currently the CSS file is served via the CDN so requests to `/assets` are relative to the CDN, not this repo.)

I really like your point about removing the `/developers` part from this file and using the `baseurl` instead. That's much more flexible. Is it alright if we leave the paths as-is for now and address that piece of it when we wire up Webpack to this repo? It'll be much easier to inject dynamic values like `baseurl` once we have proper build tools in place.

Comment (andryanmiller) [reply css/vendors.css]: Gotcha. I hadn't realized we still had an asset dependency! My take: let's make a ticket to migrate those resources out of Play and into the dev-center, and tackle it directly after this ticket. Feel to mark with `Docs App` epic and move into this sprint. I'm ok with these paths being the way they are until we address that story.

File: _layouts/default.html [hunk 5]
```diff
@@ -65,6 +66,7 @@
     <script src="{{ site.baseurl }}/js/search.min.js" type="text/javascript" charset="utf-8"></script>
     <script src="{{ site.baseurl }}/js/userCreds.js" type="text/javascript" charset="utf-8"></script>
     <script src="{{ site.baseurl }}/js/stageCustomization.js" type="text/javascript" charset="utf-8"></script>
+    <script src="https://stackpath.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js"></script>
```
Comment (andryanmiller) [inline _layouts/default.html]: ⚠️ Why are we pulling this resource from an external CDN?

File: css/vendors.css (hunk 4 above)
Comment (andryanmiller) [reply css/vendors.css]: Alright, having now absorbed all of this, things make much more sense. I do think having a `<base>` tag will address most of these woes. Mind creating a ticket for the "Docs App" epic that captures doing our best to make all relative URLs location-agnostic? I agree the webpack setup will make this easier, but I'd like to explore if there's a potentially cheap way to do it with Jekyll as well, as this is a somewhat sharp edge I'd like to sand off.

File: server/index.js [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+  } else {
+    next()
+  }
+})
+
+// Dev Center
+
+const isDirectory = path => !/\w+\.\w+$/.test(path)
+app.use(
+  /^\/developers/,
```
Comment (andryanmiller) [inline server/index.js]: :bulb: Ah, per the note in the ticket, I'd prefer if this server didn't have to know anything about the `/developers` path. Perhaps there's a technical constraint I'm missing?

**Comment by peckjon** [comment]
Dang... was really hoping this last merge would fix it, but here's my local render... bit of a problem b/c the bad header covers then whole page z-index so I can't interact with the content area @bnlb @langlk 

Screen Shot 2019-10-02 at 5 24 05 PM

**Comment by peckjon** [comment]
Slightly better using Web -- local proxy is correct for content, and header is usable, but images are still getting requested from https://cdn.algorithmia.com, and 404 for new local images

Screen Shot 2019-10-02 at 5 32 06 PM

**Comment by langlk** [comment]
@peckjon have you built the synapse files locally?([step 5 here](https://github.com/algorithmiaio/dev-center#getting-set-up))

**Comment by peckjon** [comment]
Yay!  Thanks @langlk that worked beautifully. Apologies for my failure to carefully read the new setup.

Presumably I'll need to **git submodule update && cd synapse && npm ci && npm run build && cd ..** anytime that Synapse gets updated, but not otherwise?

**Comment by langlk** [comment]
> Yay! Thanks @langlk that worked beautifully. Apologies for my failure to carefully read the new setup.
> 
> Presumably I'll need to **git submodule update && cd synapse && npm ci && npm run build && cd ..** anytime that Synapse gets updated, but not otherwise?

Correct!

**Comment by bnlb** [comment]
> @peckjon have you built the synapse files locally?([step 5 here](https://github.com/algorithmiaio/dev-center#getting-set-up))

@StephanieKim just wanted to shoot you a message about this so you're not caught off guard the next time you open this repo. Synapse is now a submodule so things may look a bit funky the next time you fire things up if you don't build the styles first.

**Comment by peckjon** [comment]
LL in order to not get the CDN problem, I need to:
  **bundle exec jekyll serve --config _config.yml,_config-dev.yml**

If that sounds right @langlk @bnlb, I'll update the instructions...

**Comment by bnlb** [comment]
If you haven't gotten to that I'm happy to put in a quick PR to update the documentation.

**Comment by peckjon** [comment]
Yes please thx @bnlb !

**Pull request merged**